### PR TITLE
Issue 705 improve pull background jobs

### LINKED ISSUE #677: Error management and parallelization in Pull and Push forms
This issue tries to document the root causes of some Sentry error reports we've been having. The following notes gather what has surfaced while studying these error reports, which involves how we deal with parallelism and errors in the pull and push operations, among other things.

- We currently use a combination of executor services, futures and callables.
- Executor services are initialized arbitrarily, as opposed to how the more up to date ForkJoinPool works, which automatically adapts to the host machine's resources.
- Information about the specific checked exceptions that the callables throw get lost because the executor services deal with the `Exception` superclass only.
- It's not clear how exceptions occurring in secondary threads from the executor services escalate to the main thread, nor how to control them before they reach it.

It's going to be hard for us to fine tune error management around pull and push in the current scenario because investing time and effort in improving what now (after Java8) has become an obsolete parallelism and error management paradigm doesn't look like the best thing to do.

Going forward, I'd propose we refactor Briefcase to:
- Move away from checked exceptions
- Use ForkJoinPool/parallel Streams for parallelism
- Use the API around CompletableFuture for error management and operation cancellation

We have already made a step in this direction regarding the export operation. Doing this on the pull and push operations feels like the right thing to do.

### LINKED ISSUE #705: Improve pull operation background jobs
This issue is intended to track the code refactoring goals regarding parallelism, use of the HTTP client and background job error management and user feedback.

### PR DESCRIPTION
**Pull request by ggalmazor**
Closes #677, #705

This PR builds on top of #698 

#### What has been done to verify that this works as intended?
Added new automated tests generally covering the pull process
Tested manually on Linux(Ubuntu) using an Aggregate v2.0 local server:
- Pull from UI & CLI
- Export with the pull before export option enabled from UI & CLI

#### Why is this the best possible solution? Were any other approaches considered?
This PR addresses some issues on parallelism, the use of the HTTP client, error management, and user feedback

This PR also fixes console user feedback.

This work is quite invasive, although the original code hasn't been changed. To address the risks of breaking things I've added several automated tests, including integration tests that focus on the Aggregate API interaction.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

- Cancelling pull operations from the UI should feel snappier
- The pull process makes better use of the CPU and network resources. Users might expect the pull operation to go smoother/faster
- Users should expect to see the whole output of the pull process now.
- **The "pull in parallel" has no effect whatsoever** 
  - Briefcase will delegate on the JVM to use all the local computing power available, which leaves the bottlenecks on the filesystem IO, and HTTP IO
  - The `http.maxConnections` system property controls how many simultaneous connections can be created (defaults to 5). 
  - Apache Commons HttpClient will allow `http.maxConnections` max connections per route, and `http.maxConnections` * 2 max total connections.

#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.
Nope.

### CODE REVIEW

**Comment by ggalmazor** [comment]
@dcbriccetti, I'm tagging you to review this one, but we need to focus on #698 first ;)

**Comment by ggalmazor** [comment]
I'm redoing the git commit story to make this PR easier to review

**Comment by ggalmazor** [comment]
Now there are more commits but they should be smaller and have fewer concept changes. I hope that it makes the review easier! :)

**Comment by dcbriccetti** [comment]
> I hope that it makes the review easier! :)
We’ll know soon! :-)

File: src/org/opendatakit/briefcase/pull/PullTracker.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  void trackFormDownloaded() {
+    form.setStatusString("Downloaded form " + form.getFormName());
+    log.info("Downloaded form {}", form.getFormName());
+    notifyTrackingEvent();
+  }
+
+  void trackBatches(List<InstanceIdBatch> batches) {
+    totalSubmissions = batches.stream().map(InstanceIdBatch::count).reduce(0, (a, b) -> a + b);
```
[the remark this answers is not in the document]
Comment (ggalmazor) [reply src/org/opendatakit/briefcase/pull/PullTracker.java]: Yes, thanks! This is already taken care of in an upcoming PR

File: src/org/opendatakit/briefcase/reused/job/JobsRunner.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  private final ExecutorService executor;
+
+  public JobsRunner() {
+    executor = new ForkJoinPool(
+        ForkJoinPool.commonPool().getParallelism(),
+        ForkJoinPool.commonPool().getFactory(),
+        (thread, throwable) -> errorCallbacks.forEach(c -> c.accept(throwable)),
+        ForkJoinPool.commonPool().getAsyncMode()
+    );
+  }
```
[the remark this answers is not in the document]
Comment (ggalmazor) [reply src/org/opendatakit/briefcase/reused/job/JobsRunner.java]: You're probably right. Let's try to remember about this for upcoming PRs :)

**Pull request merged**